File: android/src/main/kotlin/com/emarsys/emarsys_sdk/mapper/RecommendationFilterListMapper.kt
package com.emarsys.emarsys_sdk.mapper

import com.emarsys.predict.api.model.RecommendationFilter

class RecommendationFilterListMapper : Mapper<List<Map<String, Any?>>, List<RecommendationFilter>> {
    override fun map(input: List<Map<String, Any?>>): List<RecommendationFilter> {
        return input.map { filterMap ->
            mapRecommendationFilter(filterMap)
        }.filterNotNull()
    }

    private fun mapRecommendationFilter(input: Map<String, Any?>): RecommendationFilter? {
        val expectations = input["values"] as List<String>
        if (input["filterType"] == "INCLUDE") {
            val filter = RecommendationFilter.include(input["field"] as String)
            return when (input["comparison"]) {
                "IS" -> filter.isValue(expectations[0])
                "HAS" -> filter.hasValue(expectations[0])
                "IN" -> filter.inValues(expectations)
                "OVERLAPS" -> filter.overlapsValues(expectations)
                else -> null
            }
        } else if (input["filterType"] == "EXCLUDE") {
            val filter = RecommendationFilter.exclude(input["field"] as String)
            return when (input["comparison"]) {
                "IS" -> filter.isValue(expectations[0])
                "HAS" -> filter.hasValue(expectations[0])
                "IN" -> filter.inValues(expectations)
                "OVERLAPS" -> filter.overlapsValues(expectations)
                else -> null
            }
        } else {
            return null
        }
    }
}

File: lib/mappers/recommendation_filter_list_mapper.dart
import 'package:emarsys_sdk/emarsys_sdk.dart';
import 'package:emarsys_sdk/mappers/mapper.dart';

class RecommendationFilterListMapper
    extends Mapper<List<RecommendationFilter>, List<Map<String, dynamic>>> {
  @override
  List<Map<String, dynamic>> map(List<RecommendationFilter> input) {
    return input
        .map((recommendationFilter) =>
            mapRecommendationFilter(recommendationFilter))
        .toList();
  }

  Map<String, dynamic> mapRecommendationFilter(RecommendationFilter input) {
    return {
      "filterType": input.filterType.toShortString(),
      "field": input.field,
      "comparison": input.comparison.toShortString(),
      "values": input.values
    };
  }
}

extension ParseToString on Enum {
  String toShortString() {
    return toString().split('.').last;
  }
}
